pytorch_lightning/profiler/profilers.py

```diff
@@ -543,3 +548,251 @@ def __del__(self):
+
+    def step(self, prev_action, current_action):
+        """
+        Signals the profiler that the next profiling step has started.
+        """
+        if self.record_steps and self.step_rec_fn:
+            self.step_rec_fn.__exit__(None, None, None)
+
+        if current_action == ProfilerAction.NONE:
+            if prev_action == ProfilerAction.NONE:
+                pass
+            elif prev_action == ProfilerAction.WARMUP:
+                rank_zero_warn("Incorrect schedule: WARMUP followed by NONE")
+                self._start_trace()
+                self._stop_trace()
+            elif prev_action == ProfilerAction.RECORD:
+                rank_zero_warn("Incorrect schedule: RECORD followed by NONE")
+                self._stop_trace()
+            else:
+                assert prev_action == ProfilerAction.RECORD_AND_SAVE
+                # self._stop_trace() this breaks
+                if self.on_trace_ready:
+                    self.on_trace_ready(self)
+        elif current_action == ProfilerAction.WARMUP:
+            if prev_action == ProfilerAction.NONE:
+                self._start_warmup()
+            elif prev_action == ProfilerAction.WARMUP:
+                pass
+            elif prev_action == ProfilerAction.RECORD:
+                rank_zero_warn("Incorrect schedule: RECORD followed by WARMUP")
+                self._stop_trace()
+            else:
+                assert prev_action == ProfilerAction.RECORD_AND_SAVE
+                self._stop_trace()
+                if self.on_trace_ready:
+                    self.on_trace_ready(self)
+                self._start_warmup()
+        elif current_action in \
+                [ProfilerAction.RECORD, ProfilerAction.RECORD_AND_SAVE]:
+            if prev_action == ProfilerAction.NONE:
+                self._start_warmup()
+                self._start_trace()
+            elif prev_action == ProfilerAction.WARMUP:
+                # self._start_trace() this breaks
+                pass
+            elif prev_action == ProfilerAction.RECORD:
+                pass
+            else:
+                assert prev_action == ProfilerAction.RECORD_AND_SAVE
+                self._stop_trace()
+                if self.on_trace_ready:
+                    self.on_trace_ready(self)
+                self._start_warmup()
+                self._start_trace()
+
+
+class ScheduleWrapper:
+
+    def __init__(self, schedule: Optional[Callable]):
+        self._schedule = schedule
+        self._num_step = 0
+        # used to stop profiler when `ProfilerAction.RECORD_AND_SAVE` is reached.
+        self._none_action = None
+
+    @property
+    def current_action(self):
+        if self._none_action is not None:
+            return self._none_action
+        return self._schedule(self._num_step)
+
+    @property
+    def previous_action(self):
+        if self._none_action is not None:
+            return self._none_action
+        return self._schedule(self._num_step - 1)
```

Comment:
Would it not be easier to instead have current_action and next_action?

    def current_action(self):
        if self._none_action is not None:
            return self._none_action
        return self._schedule(self._num_step)

    @property
    def next_action(self):
        if self._none_action is not None:
            return self._none_action
        return self._schedule(self._num_step + 1)

    def __call__(self, num_step:int):
        '''
        # ignore the provided input. Keep internal state instead.
        if self._none_action is not None:
            return self._none_action
        self._num_step += 1
        '''
        # if previous action is RECORD_AND_SAVE, then 
        if self._schedule(self._num_step - 1) == ProfilerAction.RECORD_AND_SAVE:
            # self._none_action = ProfilerAction.NONE
            return ProfilerAction.NONE
        return self._schedule(self._num_step)

Reply:
Remove !